File: src/com/timepath/launcher/ui/web/ProxyHandler.java
package com.timepath.launcher.ui.web;

import com.sun.net.httpserver.Headers;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;

/**
 *
 * @author TimePath
 */
class ProxyHandler implements HttpHandler {

    public void handle(HttpExchange t) throws IOException {
        String loc = t.getRequestURI().toString().substring(Server.ENDPOINT_PROXY.length() + 1);
        System.out.println("Proxy: " + loc);
        URL url = new URL(loc);
        Headers headIn = t.getRequestHeaders();
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestProperty("User-Agent", headIn.get("User-Agent").get(0));
        conn.connect();
        OutputStream os = t.getResponseBody();
        int code = conn.getResponseCode();
        int broad = code / 100;
        switch(broad) {
            case 2:
                String len = conn.getHeaderField("Content-Length");
                long size = -1;
                try {
                    size = Long.parseLong(len);
                } catch(NumberFormatException e) {
                }
                t.sendResponseHeaders(code, size); // TODO: return code
                byte[] buffer = new byte[8192];
                InputStream is = new BufferedInputStream(conn.getInputStream(), buffer.length);
                int read;
                while((read = is.read(buffer)) > -1) {
                    os.write(buffer, 0, read);
                    os.flush();
                }
                break;
            case 3:
                System.out.println("redirect: " + conn.getHeaderField("Location"));
                break;
            default:
                System.out.println("err: " + code);
                break;
        }
        os.close();
    }
    
}

File: src/main/kotlin/com/timepath/launcher/ui/web/ProxyHandler.kt
package com.timepath.launcher.ui.web

import com.sun.net.httpserver.HttpExchange

import java.io.*
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.logging.Level
import java.util.logging.Logger

/**
 * @author TimePath
 */
class ProxyHandler {
    class object {

        private val LOG = Logger.getLogger(javaClass<ProxyHandler>().getName())

        /**
         * Handles page requests of the form: 'http://www.something.com'
         *
         * @param t
         * @param loc
         */
        public fun handleProxy(t: HttpExchange, loc: String) {
            LOG.log(Level.INFO, "Proxy: {0}", loc)
            val url: URL
            try {
                url = URL(loc)
            } catch (ex: MalformedURLException) {
                LOG.log(Level.SEVERE, null, ex)
                return
            }

            val headIn = t.getRequestHeaders()
            val conn: HttpURLConnection
            try {
                conn = url.openConnection() as HttpURLConnection
                conn.setRequestProperty("User-Agent", headIn["User-Agent"][0])
                conn.setRequestMethod(t.getRequestMethod())
                conn.connect()
            } catch (ex: IOException) {
                LOG.log(Level.SEVERE, null, ex)
                return
            }

            try {
                t.getResponseBody().use { os ->
                    val code = conn.getResponseCode()
                    val broad = code / 100
                    when (broad) {
                        2 -> {
                            val size = conn.getContentLengthLong()
                            val baos = ByteArrayOutputStream(size.toInt())
                            conn.getInputStream().buffered().copyTo(baos)
                            val doc = String(baos.toByteArray(), StandardCharsets.UTF_8)
                            val cType = conn.getContentType()
                            val raw = doc.toByteArray()
                            val responseHeaders = t.getResponseHeaders()
                            responseHeaders.set("Content-Type", cType)
                            t.sendResponseHeaders(code, raw.size().toLong()) // TODO: proper return code handling
                            os.write(raw)
                        }
                        3 -> {
                            LOG.log(Level.INFO, "{0} -> {1}", array<Any>(code, conn.getHeaderField("Location")))
                            t.sendResponseHeaders(code, 0) // TODO: redirect
                        }
                        else -> {
                            LOG.log(Level.INFO, "{0}: {1}", array<Any>(code, loc))
                            t.sendResponseHeaders(code, 0)
                        }
                    }
                }
            } catch (ex: IOException) {
                LOG.log(Level.SEVERE, null, ex)
            }

        }

        throws(javaClass<MalformedURLException>())
        fun checkProxy(exchange: HttpExchange, request: String): Boolean {
            var proxyRequest: String? = null
            if (request.startsWith(Server.ENDPOINT_PROXY)) {
                proxyRequest = request
            } else {
                val ref = exchange.getRequestHeaders().getFirst("Referer")
                LOG.log(Level.FINE, "Has referer {0}", ref)
                if (ref != null) {
                    val path = URL(ref + '/' + request).getPath()
                    if (path.startsWith(Server.ENDPOINT_PROXY)) {
                        proxyRequest = path
                    }
                }
            }
            if (proxyRequest != null) {
                handleProxy(exchange, proxyRequest!!.substring(Server.ENDPOINT_PROXY.length() + 1)) // Remove leading '/proxy/'
                return true
            }
            return false
        }
    }
}
